### PR TITLE
Create lib to fetch feature switches and store in DB

### PR DESCRIPTION
**Pull request by vitaliyreznikov**
Create lib to fetch feature switches and store in DB

Fixes: https://app.asana.com/0/1163213880723410/1163214272245398/f

#changelog Create lib to fetch feature switches and store in DB.

# Type of change
- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Test

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [x] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have created or modified unit tests to sufficiently cover my changes.
- [x] I have added my Asana task link and provided a changelog message if applicable.

### CODE REVIEW

File: app/code/community/Bolt/Boltpay/Helper/FeatureSwitch/ManagerTrait.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+        $switches = array();
+        foreach ($switchesData as $switch) {
+            $switches[$switch->name] = (object)array(
+                'value' => $switch->value,
+                'defaultValue' => $switch->defaultValue,
+                'rolloutPercentage' => $switch->rolloutPercentage
+            );
+        }
+        Mage::getModel('core/config')->saveConfig('payment/boltpay/feautureSwitches', json_encode($switches));
```
Comment (matt-thomason) [inline app/code/community/Bolt/Boltpay/Helper/FeatureSwitch/ManagerTrait.php]: Typo here - should be `payment/boltpay/featureSwitches`.
Has this table been set up already? I don't see an upgrade script to add it.

Comment (adentong) [reply app/code/community/Bolt/Boltpay/Helper/FeatureSwitch/ManagerTrait.php]: Good catch.

Comment (vitaliyreznikov) [reply app/code/community/Bolt/Boltpay/Helper/FeatureSwitch/ManagerTrait.php]: Thank you @matt-thomason. Fixed

File: app/code/community/Bolt/Boltpay/Helper/FeatureSwitch/ManagerTrait.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            $switches[$switch->name] = (object)array(
+                'value' => $switch->value,
+                'defaultValue' => $switch->defaultValue,
+                'rolloutPercentage' => $switch->rolloutPercentage
+            );
+        }
+        Mage::getModel('core/config')->saveConfig('payment/boltpay/featureSwitches', json_encode($switches));
+        Mage::getModel('core/config')->cleanCache();
+    }
+}
```
Comment (mccottry) [inline app/code/community/Bolt/Boltpay/Helper/FeatureSwitch/ManagerTrait.php]: For me, this is not a Helper, it is a model, especially when we are using it to persist to the database. Also, there is no need to over-complicate things with a FeatureSwitch directory and several class.  Everything that we are going to need here can go into `Bolt_Boltpay_Model_FeatureSwitch`.  

@johnmccombs1 How do you feel about this?

Comment (vitaliyreznikov) [reply app/code/community/Bolt/Boltpay/Helper/FeatureSwitch/ManagerTrait.php]: It makes sense Leon. @johnmccombs1 waiting for your approval before moving this to model

Comment (mccottry) [reply app/code/community/Bolt/Boltpay/Helper/FeatureSwitch/ManagerTrait.php]: @vitaliyreznikov You can probably move forward with this simple change request.

File: app/code/community/Bolt/Boltpay/Model/FeatureSwitch.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+class Bolt_Boltpay_Model_FeatureSwitch extends Bolt_Boltpay_Model_Abstract
+{
+    /**
+     * This method gets feature switches from Bolt and updates the local DB with
+     * the latest values. To be used in upgrade data and webhooks.
+     */
+    public function updateSwitchesFromBolt()
+    {
+        $switchesResponse = $this->boltHelper()->getFeatureSwitches();
+        $switchesData = @$switchesResponse->data->plugin->features;
```
Comment (roopakv) [inline app/code/community/Bolt/Boltpay/Model/FeatureSwitch.php]: is the `@` operator supported on all versons of PHP we support?

File: app/code/community/Bolt/Boltpay/Model/FeatureSwitch.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+        $switches = array();
+        foreach ($switchesData as $switch) {
+            $switches[$switch->name] = (object)array(
+                'value' => $switch->value,
+                'defaultValue' => $switch->defaultValue,
+                'rolloutPercentage' => $switch->rolloutPercentage
+            );
+        }
+        Mage::getModel('core/config')->saveConfig('payment/boltpay/featureSwitches', json_encode($switches));
```
Comment (roopakv) [inline app/code/community/Bolt/Boltpay/Model/FeatureSwitch.php]: why not `payment/boltpay/featureSwitches/$switchName` that way reading is quicker? si config explosion not good? i cant find any good guidelines

File: app/code/community/Bolt/Boltpay/Model/FeatureSwitch.php (hunk 3 above)
Comment (mccottry) [reply app/code/community/Bolt/Boltpay/Model/FeatureSwitch.php]: Yes it is supported on all versions of PHP that we support. 

The `@` operator, in this context has been supported for at least the last 13 years and will likely be supported for the foreseeable future of PHP.

File: app/code/community/Bolt/Boltpay/Model/FeatureSwitch.php (hunk 4 above)
Comment (mccottry) [reply app/code/community/Bolt/Boltpay/Model/FeatureSwitch.php]: As I interpret this approach, the updating to the database and reading into memory of all switches is done once... or at least it is set up for that pattern.  It seems perfectly natural for periodic bulk updates on software upgrade or push request to update.  Regarding reading a particular feature, one single call to the database and then reading from memory seems faster to me than separate calls to the database for each feature.  Memory footprint is probably negligible in this use case.

Comment (vitaliyreznikov) [reply app/code/community/Bolt/Boltpay/Model/FeatureSwitch.php]: Yes, I agree with Leon. It's better to have only one SQL call. Until we have no more than a few dozens of feature switch this approach should work much faster

Comment (adentong) [reply app/code/community/Bolt/Boltpay/Model/FeatureSwitch.php]: @roopakv Do you still feel strongly about this? I'd like to get your approval on this PR so it doesn't sit there in an almost ready state waiting for one last comment to be resolved.

Comment (roopakv) [reply app/code/community/Bolt/Boltpay/Model/FeatureSwitch.php]: Nope was a question. I agree with Leon and vitality

**Pull request merged**